pages/products-details-page.ts

```diff
@@ -0,0 +1,122 @@
+    }
+
+    @boxedStep
+    private async openReviewsTab() {
+        while( ! await this.reviewsTabOpen() ) { 
+            await this.page.locator('#tab-label-reviews').click();
+        }
+    }
+
+    private async reviewsTabOpen(): Promise<boolean> {
+        // TODO: extract to fields
+        return this.tabsContainer.getByTestId('nickname_field').isVisible();
+    }
+
+    @boxedStep
+    async addReview(review: Review) {
+        await this.openReviewsTab();
+        const ratingLabel = this.page.getByTestId(ProductDetailsPage.getTestId(review.rating));
+        const ratingLabelBoundingBox = await ratingLabel.boundingBox()
+            .then( box => throwOnNull(box));
+
+        // click on star corresponding to rating
+        const ratingClickPosition = {
+            x: ratingLabelBoundingBox.width - ratingLabelBoundingBox.height/2,
+            y: ratingLabelBoundingBox.height/2 
+        }
+        await ratingLabel.click({position: ratingClickPosition});
+
+        // firefox seems to clear fields for some reason
+        await expect( async () => {
+            await this.tabsContainer.getByTestId('nickname_field').fill(review.nickname);
+            await this.tabsContainer.getByTestId('summary_field').fill(review.summary);
+            await this.tabsContainer.getByTestId('review_field').fill(review.content);
+            //expect(await this.areReviewFieldsFilled()).toBeTruthy();
+            await this.tabsContainer.locator('button[type="submit"]').click()
+            // make sure that there is no error messages related to fields
+            await expect(this.tabsContainer.locator(".mage-error")).not.toBeAttached()
+            .catch( (reason) => {
+                log('Repeating, because: ' + reason);
+                throw new Error(reason);
+            })
+        }).toPass();
+    }
+    
+    @boxedStep
+    async getProductData(): Promise<Product> {
+        const name = (await this.productNameLocator.textContent())?.trim() ?? '';
+        const price = await this.priceLocator.textContent().then(parsePrice)
+        let ratingsCount: number;
+        let avgRating: number;
+
+        if( await this.nonEmptyRatingsContainer.isVisible()) {
+            ratingsCount = parseRatingsCount(await this.ratingsCountLocator.textContent());
+            avgRating = await this.avgRatingLocator.getAttribute('title')
+                .then(s => Number.parseFloat(s ?? 'NaN'));
+        } else {
+            ratingsCount = 0;
+            avgRating = NaN;
+        }
+        return new Product(name, price, ratingsCount, avgRating);
+    }
+
+    private static getTestId(rating: Review.Rating) {
+        return 'Rating_' + rating  + '_label';
+    }
+
+    // TODO: make this into type
+    // FIXME: make this return {content: string, status: 'success' | 'notice' | 'error'} type
+    @boxedStep
+    private async getMessages() :Promise<{content: string, status: string}[]> {
```

Comment:
If you need to return multiple values from a method, unite these values to a object